File: node_version/generate.ts
import { getActiveProvider } from "./providers/registry.js"

export async function generateExplanation(
  prompt: string,
  providerOverride?: string
): Promise<string> {

  const provider = getActiveProvider(providerOverride)

  try {

    const output = await provider.generate(prompt)

    if (!output || !output.trim()) {
      throw new Error(`${provider.name} returned no output`)
    }

    return output.trim()

  } catch (err: any) {

    const message =
      err?.message ? String(err.message) : String(err)

    throw new Error(
      `${provider.name} generation failed: ${message}`
    )

  }

}

File: explain_this_repo/generate.py
from explain_this_repo.providers.registry import get_active_provider


def generate_explanation(
    prompt: str,
    provider_override: str | None = None,
) -> str:
    provider = get_active_provider(override=provider_override)

    try:
        output = provider.generate(prompt)
    except Exception as e:
        raise RuntimeError(f"{provider.name} generation failed: {e}") from e

    if not output or not output.strip():
        raise RuntimeError(f"{provider.name} returned no output")

    return output.strip()
